source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -50,16 +53,37 @@ LocalRateLimiterImpl::LocalRateLimiterImpl(
         PROTOBUF_GET_WRAPPED_OR_DEFAULT(descriptor.token_bucket(), tokens_per_fill, 1);
     new_descriptor.token_bucket_ = token_bucket;
 
-    auto token_state = std::make_unique<TokenState>();
+    auto token_state = std::make_shared<TokenState>();
     token_state->tokens_ = token_bucket.max_tokens_;
     token_state->fill_time_ = time_source_.monotonicTime();
-    new_descriptor.token_state_ = std::move(token_state);
+    new_descriptor.token_state_ = token_state;
 
-    auto result = descriptors_.emplace(std::move(new_descriptor));
+    auto result = descriptors_.emplace(new_descriptor);
     if (!result.second) {
       throw EnvoyException(absl::StrCat("duplicate descriptor in the local rate descriptor: ",
                                         result.first->toString()));
     }
+    sorted_descriptors_.push_back(new_descriptor);
+  }
+  // If a request is limited by a descriptor, it should not consume tokens from the remaining
+  // matched descriptors, so we sort the descriptors by tokens per second, as a result, in most
+  // cases the strictest descriptor will be consumed first. However, it can not solve the
+  // problem perfectly.
+  if (!sorted_descriptors_.empty()) {
+    std::sort(sorted_descriptors_.begin(), sorted_descriptors_.end(),
+              [](const LocalDescriptorImpl a, const LocalDescriptorImpl b) -> bool {
+                const int a_token_rate_per_second =
+                    a.token_bucket_.tokens_per_fill_ /
+                    (absl::ToInt64Seconds(a.token_bucket_.fill_interval_)
+                         ? absl::ToInt64Seconds(a.token_bucket_.fill_interval_)
+                         : 1);
```

Comment:
nit: to make this easier to read can you factor out the per second calculation to a lambda or different function?